## Intro to Mod3 Assessment 

In the cells below, we are importing relevant modules that you might need later on. We also load and prepare the dataset for you.

In [1]:
# Run this cell without changes
import pandas as pd
import itertools
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np
from sklearn.linear_model import Lasso, Ridge
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures

In [2]:
# Run this cell without changes
data = pd.read_csv('Advertising.csv').drop('Unnamed: 0',axis=1)
data.describe()

TV       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   30.554000   14.022500
std     85.854236   14.846809   21.778621    5.217457
min      0.700000    0.000000    0.300000    1.600000
25%     74.375000    9.975000   12.750000   10.375000
50%    149.750000   22.900000   25.750000   12.900000
75%    218.825000   36.525000   45.100000   17.400000
max    296.400000   49.600000  114.000000   27.000000

**1. Drop the 'sales' column, use the correct axis and set this line of code = to X then set the 'sales' column = y** 

In [4]:
X = data.drop('Sales', axis=1)
y = data['Sales']

**2.Split the data into training and test sets and set the random state to 2019**

In [5]:
X_train , X_test, y_train, y_test = train_test_split(X, y,random_state=2019)

**3. Write a function to calculate train and test error for different polynomial degrees.**

This function should:

- take degree as a parameter that will be used to create polynomial features to be used in a linear regression model

- create a PolynomialFeatures object for each degree and fit a linear regression model using the transformed data

- calculate the mean square error for each level of polynomial

- return the train_error and test_error

In [6]:
def polynomial_regression(degree):
    """
    Calculate train and test errorfor a linear regression with polynomial features.
    (Hint: use PolynomialFeatures)
    
    input: Polynomial degree
    output: Mean squared error for train and test set
    """
    poly = PolynomialFeatures(degree=degree)
    X_poly_train = poly.fit_transform(X_train)
    X_poly_test = poly.transform(X_test)
    lr_poly = LinearRegression()
    lr_poly.fit(X_poly_train,y_train)
    train_error = mean_squared_error(y_train, lr_poly.predict(X_poly_train))
    test_error = mean_squared_error(y_test, lr_poly.predict(X_poly_test))
    return train_error, test_error

**Try out your new function below**

In [14]:
# Run this cell without changes
polynomial_regression(1)

(2.666850766486441, 3.3855752460622908)

In [7]:
# Run this cell without changes
polynomial_regression(4)

(0.18179110221108896, 1.9522601290515889)

**Check your answers**

Approximate MSE for degree 3:

Train: 0.242<br/>
Test: 0.153

Approximate MSE for degree 4:

Train: 0.182<br/>
Test: 1.95

## Gradient Descent 
![](Gradient_Descent.png)

**Q1:**Using the gradient descent visual from above, explain why the distance between each step is getting smaller as more steps occur with gradient descent.

**Answer:**
The distance between the steps is getting smaller because the slope gradually becomes less and less steep as you get closer to finding the minimum.

**Q2:**What is the purpose of a learning rate in gradient descent? Explain how a very small and a very large learning rate would affect the gradient descent.

**Answer:**
Learning rate is a number that is multiplied by each step that is taken during gradient descent. If the learning rate is smaller, the step sizes will become smaller. If the learning rate is larger, the step sizes will be larger. Learning rate is present in gradient descent to help ensure that an optimal minimum on the cost curve is discovered.

## Confusion Matices 
![](confusion_matrix.png)

**Q1.** Using the confusion matrix up above, calculate precision, recall, and F-1 score.

In [8]:
#Calculate using python here
precision = 30/(30+4)
recall = 30 / (30 + 12)
F1 = 2 * (precision * recall) / (precision + recall)

print("precision: {}".format(precision))
print("recall: {}".format(recall))
print("F1: {}".format(F1))

precision: 0.8823529411764706
recall: 0.7142857142857143
F1: 0.7894736842105262


**Q2.** What is a real life example of when you would care more about recall than precision? Make sure to include information about errors in your explanation.

We would care more about recall than precision in cases where a Type II error (a False Negative) would have serious consequences. An example of this would be a medical test that determines if someone has a serious disease. A higher recall would mean that we would have a higher chance of identifying all people who ACTUALLY had the serious disease.

 

**Q3.** What is an example of when you would care more about precision than recall? 

Precision is more important than recall when you would like to have less False Positives and don't mind having more False Negatives. Meaning, getting a False Positive is very costly, and a False Negative is not as much.

In a zombie apocalypse, you would try to accept as many as healthy people you can into your safe zone, but you really dont want to mistakenly pass a zombie into the safe zone. So if your method causes some of the healthy people mistakenly not to get into the safe zone, then so be it.